radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/service/EngineStatusService.java

```diff
@@ -0,0 +1,203 @@
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.api.service;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.inject.Inject;
+import com.radixdlt.consensus.LedgerProof;
+import com.radixdlt.engine.RadixEngine;
+import com.radixdlt.engine.RadixEngineReader;
+import com.radixdlt.statecomputer.LedgerAndBFTProof;
+import com.radixdlt.statecomputer.forks.CandidateForkVote;
+import com.radixdlt.statecomputer.forks.ForkVotingResult;
+import com.radixdlt.statecomputer.forks.Forks;
+import com.radixdlt.statecomputer.forks.ForksEpochStore;
+import com.radixdlt.store.LastEpochProof;
+import com.radixdlt.store.LastProof;
+import com.radixdlt.sync.CommittedReader;
+import java.util.Optional;
+
+public final class EngineStatusService {
+  private final RadixEngine<LedgerAndBFTProof> radixEngine;
+  private final CommittedReader committedReader;
+  private final LedgerProof lastProof;
+  private final LedgerProof lastEpochProof;
+  private final Forks forks;
+  private final ForksEpochStore forksEpochStore;
+
+  private long cachedForksValuesLastEpoch = -1;
+  private short cachedCandidateForkVotingResult = 0;
+  private Optional<Long> cachedUpcomingForkRemainingEpochs = Optional.empty();
+
+  @Inject
+  public EngineStatusService(
+      RadixEngine<LedgerAndBFTProof> radixEngine,
+      CommittedReader committedReader,
+      @LastProof LedgerProof lastProof,
+      @LastEpochProof LedgerProof lastEpochProof,
+      Forks forks,
+      ForksEpochStore forksEpochStore) {
+    this.radixEngine = radixEngine;
+    this.committedReader = committedReader;
+    this.lastProof = lastProof;
+    this.lastEpochProof = lastEpochProof;
+    this.forks = forks;
+    this.forksEpochStore = forksEpochStore;
+
+    refreshForksCachedValues();
+  }
+
+  public LedgerProof getEpochProof(long epoch) {
+    return committedReader.getEpochProof(epoch).orElse(lastEpochProof);
+  }
+
+  public LedgerProof getCurrentProof() {
+    var ledgerAndBFTProof = radixEngine.read(RadixEngineReader::getMetadata);
+    return ledgerAndBFTProof == null ? lastProof : ledgerAndBFTProof.getProof();
+  }
+
+  public Optional<Long> getCandidateForkRemainingEpochs() {
+    refreshForksCachedValues();
+    return this.cachedUpcomingForkRemainingEpochs;
+  }
+
+  private void refreshForksCachedValues() {
```

Comment:
A couple of things:
* There's a race condition about getCurrentProof changing during calculations; which would result in the cached values being against the wrong epoch - I think it would be better to pass the current epoch into the calculation methods
* There's no locking here, so multiple simultaneous calls could make lots of calculation calls - we should probably lock around the calculation.

Reply:
Addressed both.